qa/pull-tester/rpc-tests.py

```diff
@@ -378,7 +379,7 @@ def get_next(self):
             log_stderr = tempfile.SpooledTemporaryFile(max_size=2**16)
             self.jobs.append((t,
                               time.time(),
-                              subprocess.Popen((self.tests_dir + t).split() + self.flags + port_seed,
+                              subprocess.Popen((["shard.py"] if os.environ["SHARD_SCRIPT"] == "true" else []) + (self.tests_dir + t).split() + self.flags + port_seed,
```

Comment:
What does `shard.py` do, and where does it come from? Is there code that generates that? It would be helpful to have a comment here identifying how it's created and maybe a link to the related code.

Reply:
As mentioned above, custom_script is shard.py and they are doing essentially  https://github.com/Electric-Coin-Company/k8s-helm-templates/blob/1d5fe06c4d1e2661a7b93ee70ad357209f9598b0/templates/tekton/pipelines/zcash/tasks/zcash-rpc-test.yml#L51-L126 .

Reply:
This URL is broken for me, I presume because I don't have permissions on that repository. So I cannot review it.

Reply:
Why can't the content of `shard.py` just be in the PR? I know it's slightly specific to our CI setup, but so is the rest of `qa/pull-tester`.

Reply:
```
        #!/bin/bash
        set -e -o pipefail
        echo "#################"
        echo "# Starting task #"
        echo "#################"
        cp -a /home/.zcash-params $HOME/
        cat <<EOF > /script.sh
        #!/bin/bash
        export exit_code="0"
        export random=\$(tr -dc a-z </dev/urandom | head -c 8 ; echo '')
        export pipelinerun_uid=$(params.pipelineRun-uid)
        cat <<EOD > \$random.yaml
        apiVersion: batch/v1
        kind: Job
        metadata:
          name: \$random
          namespace: pipelines
        spec:
          ttlSecondsAfterFinished: 600
          backoffLimit: 10
          template:
            metadata:
              name: \$random
            spec:
              initContainers:
              - name: copy-artifact
                image: google/cloud-sdk
                volumeMounts:
                - name: workspace
                  mountPath: /workspace
                command:
                  - bash
                args:
                  - -c
                  - |
                    gsutil -m cp gs://\$(gcloud config list --format 'value(core.project)' 2>/dev/null)-tekton-artifacts/$(params.pipelineRun-uid)-rpc.tar /workspace/
                    cd /workspace
                    tar -xvf $(params.pipelineRun-uid)-rpc.tar
              containers:
              - name: rpc
                image: $(params.BUILD_CONTAINER)
                resources:
                  limits:
                    memory: "4G"
                    cpu: "4"
                  requests:
                      memory: "1G"
                      cpu: "0.7"
                env:
                - name: ZCASHD
                  value: /workspace/$(params.pipelineRun-uid)/source/src/zcashd
                command:
                  - bash
                  - "-c"
                  - |
                    cp -a /home/.zcash-params /root/
                    \$@
                volumeMounts:
                - name: workspace
                  mountPath: /workspace
              serviceAccount: sa-github-gcr-gcs
              serviceAccountName: sa-github-gcr-gcs
              volumes:
                - name: workspace
                  emptyDir: {}
              restartPolicy: OnFailure
              nodeSelector:
                node_pool: intensive-node-pool-c2
              affinity:
                nodeAffinity:
                  requiredDuringSchedulingIgnoredDuringExecution:
                    nodeSelectorTerms:
                      - matchExpressions:
                          - key: node_pool
                            operator: In
                            values:
                              - intensive-node-pool-c2
              tolerations:
                - key: intensive-node-pool-c2
                  operator: Exists
                  effect: NoSchedule
        EOD
        # wait for completion
        (kubectl apply -f \$random.yaml &> /dev/null && kubectl wait --for=condition=complete --timeout=3000s job/\$random -n pipelines &> /dev/null) || export exit_code="1"
        if [[ "\$exit_code" -eq "1" ]]; then
          export exit_code="0"
          (kubectl apply -f \$random.yaml &> /dev/null && kubectl wait --for=condition=complete --timeout=3000s job/\$random -n pipelines &> /dev/null) || export exit_code="1"
        fi
        if [[ "\$exit_code" -eq "1" ]]; then
          kubectl logs job/\$random -n pipelines -c rpc || echo "";
          kubectl delete job/\$random -n pipelines &> /dev/null || echo "";
          exit 1
        else
          kubectl delete job/\$random -n pipelines &> /dev/null || echo "";
          exit 0
        fi
        EOF
        chmod +x /script.sh
        HOST=$(params.BUILD_FOR_HOST) CUSTOM_SCRIPT=/script.sh /workspace/$(params.pipelineRun-uid)/source/qa/pull-tester/rpc-tests.py -m-1 -r-1 -j150
```

In the above, that is called by shard.py, the large machine with `j150` uses fraction of CPU (e.g. 0.7 instead of 1.0) to create a pod per RPC test. This reduces the net overhead of spinning up a separate pod per test, outside of a single huge machine. Each pod's entrypoint is the given RPC test.

Reply:
Why isn't `shard.py` in this repo?

Reply:
In standup, @daira, @nuttycom, @sellout and I tried to figure out what the above logic is doing. Here is our best guess:

Tekton YAML (where we believe this is embedded):
- Substitutes unescaped `$`s as Tekton config values
- Unescapes `\$`s
- Runs the resulting bash script, which:
  - Interprets `<<<EOF` as a here script
  - Starts generating `script.sh`
  - Evaluates the now-unescaped `\$` commands, and hard-codes things like a `random` value, and the URL of the parent Kubernetes controller, into `script.sh` (both outside and inside the `<<<EOD` section, so hard-coding into both `script.sh` and into the YAML file it will generate).
  - Calls `CUSTOM_SCRIPT=script.sh rpc-test.py` (now replaced with `shard.py` existing), which:
    - Decides how to split up the tests
    - Sets up the config args for subprocesses to run individual tests.
    - Fires off subprocesses, which get wrapped in `script.sh`, which:
      - Interprets `<<<EOD` as a here script
      - Starts generating a YAML file with a random name (chosen by the earlier Bash invocation).
      - Somehow loads the arguments from `rpc-test.py`'s subcommand into this YAML file via `$@` (unsure on escaping order here).
      - Fires off a Kubernetes job with the YAML script, which:
        - Fetches the pre-compiled `zcashd-...-rpc.tar` from the host URL hard-coded into the YAML file.
        - Runs bash with the specified arguments as a quoted string containing newlines, which:
          - Splits the provided argument to `-c` on newlines.
          - Runs each line as a command, which:
            - Runs the specified RPC test on the Kube.
        - Returns the results (somehow).

Reply:
If I'm understanding correctly, there were command argument injection bugs on lines 80, 82, 95, 128, 132, 136, 137, 140, and 145. All of the arguments containing $-substitutions should be properly double-quoted (for the bash subsitutions; I'm not sure about the Kubernetes YAML substititions).

Reply:
Note that we have security bugs in a file that is not in the PR, validating my complaint about that.